tests/Doctrine/Tests/ORM/Mapping/ClassMetadataFactoryTest.php

```diff
@@ -451,6 +454,70 @@ public function testInheritsIdGeneratorMappingFromEmbeddable()
 
         $this->assertTrue($userMetadata->isIdGeneratorIdentity());
     }
+
+    public function testPropertyOrderNoCache()
+    {
+        $cmf = $this->createPropertyOrderFactory();
+
+        /** @var ClassMetadataInfo $adminMetadata */
+        $adminMetadata = $cmf->getMetadataFor(PropertyOrder\Admin::class);
+        $this->assertSame(
+            ['id', 'name', 'group', 'email'],
+            array_keys($adminMetadata->reflFields)
+        );
+
+        /** @var ClassMetadataInfo $userMetadata */
+        $userMetadata = $cmf->getMetadataFor(PropertyOrder\User::class);
+        $this->assertSame(
+            ['id', 'name', 'group'],
+            array_keys($userMetadata->reflFields)
+        );
+    }
+
+    public function testPropertyOrderWithCache()
+    {
+        $noCacheCmf = $this->createPropertyOrderFactory();
+        $cache      = new ArrayCache();
+
+        $addCacheMetadata = function (string $className) use ($cache, $noCacheCmf) {
+            /** @var ClassMetadataInfo $metadata */
+            $metadata             = $noCacheCmf->getMetadataFor($className);
+            $metadata->reflFields = [];
+
+            $cacheSalt = '$CLASSMETADATA'; // \Doctrine\Common\Persistence\Mapping\AbstractClassMetadataFactory::$cacheSalt
+
+            $cache->save($className . $cacheSalt, $metadata);
+        };
+        $addCacheMetadata(PropertyOrder\Admin::class);
+        $addCacheMetadata(PropertyOrder\User::class);
+
+        $cmf = $this->createPropertyOrderFactory();
+        $cmf->setCacheDriver($cache);
+
+        /** @var ClassMetadataInfo $adminMetadata */
+        $adminMetadata = $cmf->getMetadataFor(PropertyOrder\Admin::class);
+        $this->assertSame(
+            ['id', 'name', 'group', 'email'],
```

Comment:
So, if I get it correctly, the problem is that the sorting of the keys changes? How much does the ORM API rely on fetching fields and reflection fields in the correct order? I think that since this is a hashmap, the sorting should not really cause trouble (fragile consumer, not metadata)